draft/Builder.php

```diff
@@ -0,0 +1,149 @@
+    {
+        $this->client = $client;
+        $this->options = $options;
+        $this->uri = $this->parseUri($uri);
+    }
+
+    /**
+     * Get the shortened URI asynchronously.
+     *
+     * @return \GuzzleHttp\Promise\PromiseInterface
+     */
+    public function async()
+    {
+        $driver = $this->getClient()->driver();
+        $promise = null;
+
+        if ($driver instanceof ToPromise) {
+            $promise = $driver->toPromise($this->uri, $this->options);
+        } else {
+            throw new UnsupportedOperationException('Async URL shortening is not supported');
+        }
+
+        return $promise->then(function ($uri) {
+            return $this->parseUri($uri);
+        });
+    }
+
+    /**
+     * Get the shortened URI.
+     *
+     * @return \Psr\Http\Message\UriInterface
+     */
+    public function get()
+    {
+        $driver = $this->getClient()->driver();
+        $uri = null;
+
+        if ($driver instanceof ToPromise) {
+            $uri = $driver->toPromise($this->uri, $this->options)->wait();
+        } else {
+            throw new UnsupportedOperationException('URL shortening is not supported');
+        }
+
+        return $this->parseUri($uri);
+    }
+
+    /**
+     * Get the client.
+     *
+     * @return \LaraCrafts\UrlShortener\Contracts\Client
+     */
+    public function getClient()
+    {
+        return $this->client;
+    }
+
+    /**
+     * Get the URI.
+     *
+     * @return \Psr\Http\Message\UriInterface
+     */
+    public function getUri()
+    {
+        return $this->uri;
+    }
+
+    /**
+     * Parse the given URI.
+     *
+     * @param \Psr\Http\Message\UriInterface|string $uri
+     * @return \Psr\Http\Message\UriInterface
+     */
+    protected function parseUri($uri)
+    {
+        if ($uri instanceof UriInterface) {
+            return $uri;
+        }
+
+        if (is_string($uri) || (is_object($uri) && method_exists($uri, '__toString'))) {
+            return new Uri((string)$uri);
+        }
+
+        throw new InvalidArgumentException('URI must be a string or UriInterface');
+    }
+
+    /**
+     * Set the suffix to shorten to.
+     *
+     * @param string $suffix
+     * @return $this
+     */
+    public function toSuffix(string $suffix)
+    {
+        $driver = $this->getClient()->driver();
+
+        if (!$driver instanceof WithSuffixes) {
+            throw new UnsupportedOperationException('Applying a custom suffix is not supported');
+        }
+
+        $driver->applySuffix($this, $suffix);
+
+        return $this;
+    }
+
+    /**
+     * Add a piece of data to the builder.
+     *
+     * @param array|string $key
+     * @param mixed|null $value
+     * @return $this
+     */
+    public function with($key, $value = null)
+    {
+        if (is_array($key)) {
+            $this->options = array_merge($this->options, $key);
+        } else {
+            $this->options[$key] = $value;
+        }
+
+        return $this;
```

Comment:
Isn't the builder supposed to be immutable?

```suggestion
        if (is_array($key)) {
            $options = array_merge($this->options, $key);
        } else {
            $options = $this->options;
            $options[$key] = $value;
        }
        
        return new Builder($this->client, $this->uri, $options);
```

Also this function could accept variadic parameters as well.

Reply:
I fashioned the builder after Laravel's own query builder which also is mutable, the reason I went with this is because it allows for "building" your query in loops very easily without having to constantly reassign a var (like you would with a collection).

About the variadic thing, the way it works right now you can assign one option like so: `$builder->with('password', 'secret')`. You can also mass assign options by passing an array (`$builder->with(['password' => 'secret'])`). If we make the function variadic you could this `$builder->with($password, $someOtherValue)` but we would no longer know what these variables are. Or is this not how you meant for it to be implemented?

Reply:
Oh, in that case my bad. I totally forget that options is a key-value pair.